Update Type Manager

Gorgious56 opened the pull request:
Re #5941

Add back the type dropdown and looking glass icon in the toolbar Clicking on a type in the manager no longer closes the popup anymore. A visual indictor is displayed on the active type
Changed the parameters icon in the individual type interface

Fix bug in "Slab from Walls" Shift + A code

Gorgious56 committed:
Update Type Manager

Add back the type dropdown and looking glass icon in the toolbar
Clicking on a type in the manager no longer closes the popup anymore.
A visual indictor is displayed on the active type
Changed the parameters icon in the individual type interface

Fix bug in "Slab from Walls" Shift + A code

theoryshaw commented:
hello old friend. ;)

theoryshaw commented:
would be the _cat's pajamas_ too, if the filter list was located in the launcher too... so it could filter down the thumbnails. :)

![image](https://github.com/user-attachments/assets/ed14b5aa-5adf-44fa-a21a-7451e54c0f40)

Gorgious56 committed:
Add search bar with simple filter function to the type manager

Gorgious56 commented:
Re #5941 
Here's a video illustrating the change. Any feedback before merging @Moult @theoryshaw @Andrej730 @steverugi ?

It takes a bit more space in the toolbar than before but I felt just the type manager icon wasn't visible enough.

I haven't tested the filter function on a huge database so I don't know how it fares with hundreds or thousands of types, since it triggers a bunch of redraws and recalculating of AuthoringData it might be bad. On a hundred Wall types I didn't notice any hiccup though.

![Animation3](https://github.com/user-attachments/assets/45650790-1ac1-4970-ab07-71c3798d6b9d)

Could integrate more elaborate search terms like regex or fuzzy search (might be able to use builtin Blender search functionality) afterwards

Moult commented:
Things I like:

 -  It's now clear that clicking an item makes it the active type because you immediately see the "Active" text appear.
 - The new search filter.

Things I don't like:

 - There are two ways to change the active type.
 - There are two ways to add an occurrence.
 - The interface of a dropdown, followed by a search icon, followed by "fake dropdown button thing". It looks weird. I really don't have a good solution though.

Gorgious56 committed:
Another attempt.

    Went back to previous toolbar : icon + type name + dot grid icon which all launch the type manager.
    Validating the type manager popup now only closes it. Functionally is the same thing as clicking Cancel. Don't really know what to do to fix it. Does cancel make sense in the type manager ? Do we want to be able to undo what is possible to be done from within the type manager ?
    You can change the active type either with the dropdown in the popup and in the grid flow. I changed it so that changing the type in the dropdown jumps to the corresponding page. BTW first time using github copilot to figure out the maths and it worked magically :) . I'm not against removing the dropdown but I feel like it's nice to see a plain list to choose from.
    You can input a specific page number or scrub the page number field. Might be usefull if user has dozens of pages to flip through.
    Right now the search is very simple, it does not implement fuzzy search or anything fancy with *

Gorgious56 commented:
Thanks for the feedback ! Another attempt.

- Went back to previous toolbar : icon + type name + dot grid icon which all launch the type manager.
- Validating the type manager popup now only closes it. Functionally is the same thing as clicking Cancel. Don't really know what to do to fix it. Does cancel make sense in the type manager ? Do we want to be able to undo what is possible to be done from within the type manager ?
- You can change the active type either with the dropdown in the popup and in the grid flow. I changed it so that changing the type in the dropdown jumps to the corresponding page. BTW first time using github copilot to figure out the maths and it worked magically :) . I'm not against removing the dropdown but I feel like it's nice to see a plain list to choose from.
- You can input a specific page number or scrub the page number field. Might be usefull if user has dozens of pages to flip through.
- Right now the search is very simple, it does not implement fuzzy search or anything fancy with *

I noticed a bug in the N panel with the icon generator which was caused by this PR.

Cheers

![Animation3](https://github.com/user-attachments/assets/39d53748-4aa7-45ba-a76c-7137f47d2e4f)

Gorgious56 committed:
Fix bug where the thumbnail of the active class wasn't updated

Gorgious56 committed:
Fix bug when duplicating an ifc type

Gorgious56 committed:
Fix bug when trying to add an ifc type occurence from the toolbar in the properties editor

Moult commented:
Awesome! This is the best of both worlds! A pure text dropdown, useful for those with many/complex types that use names to get around, a filter to aid it, as well as visual thumbnails all in one spot!

> Do we want to be able to undo what is possible to be done from within the type manager ?

Nah :)

The Close / Cancel is still awkward but isn't hurting anybody now. Sooner or later we'll get some inspiration of what to do with it. Feel free to merge!

Gorgious56 committed:
The filter search field now also filters description and predefined type

Gorgious56 committed:
Fix potential bug where a class doesn't have a predefined type

Gorgious56 committed:
Merge branch 'v0.8.0' of https://github.com/IfcOpenShell/IfcOpenShell into update_type_manager
